aws/rust-runtime/aws-inlineable/src/endpoint_discovery.rs

```diff
@@ -0,0 +1,287 @@
+        loop {
+            match self.rx.try_recv() {
+                Ok(_) | Err(TryRecvError::Closed) => break,
+                _ => {}
+            }
+            self.reload_increment(self.time.now()).await;
+            self.sleep.sleep(Duration::from_secs(60)).await;
+        }
+    }
+
+    async fn reload_increment(&self, now: SystemTime) {
+        let should_reload = self
+            .endpoint
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|e| e.is_expired(now))
+            .unwrap_or(true);
+        if should_reload {
+            tracing::debug!("reloading endpoint, previous endpoint was expired");
+            self.reload_once().await;
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct EndpointCache {
+    error: Arc<Mutex<Option<ResolveEndpointError>>>,
+    endpoint: Arc<Mutex<Option<ExpiringEndpoint>>>,
+    // When the sender is dropped, this allows the reload loop to stop
+    _drop_guard: Arc<Sender<()>>,
+}
+
+impl<T> ResolveEndpoint<T> for EndpointCache {
+    fn resolve_endpoint(&self, _params: &T) -> aws_smithy_http::endpoint::Result {
+        self.resolve_endpoint()
+    }
+}
+
+#[derive(Debug)]
+struct ExpiringEndpoint {
+    endpoint: Endpoint,
+    expiry: SystemTime,
+}
+
+impl ExpiringEndpoint {
+    fn is_expired(&self, now: SystemTime) -> bool {
+        tracing::debug!(expiry = ?self.expiry, now = ?now, delta = ?self.expiry.duration_since(now), "checking expiry status of endpoint");
+        match self.expiry.duration_since(now) {
+            Err(_) => true,
+            Ok(t) => t < Duration::from_secs(120),
+        }
+    }
+}
+
+pub(crate) async fn create_cache<F>(
+    loader_fn: impl Fn() -> F + Send + Sync + 'static,
+    sleep: Arc<dyn AsyncSleep>,
+    time: Arc<dyn TimeSource>,
+) -> Result<(EndpointCache, ReloadEndpoint), ResolveEndpointError>
+where
+    F: Future<Output = Result<(Endpoint, SystemTime), ResolveEndpointError>> + Send + 'static,
+{
+    let error_holder = Arc::new(Mutex::new(None));
+    let endpoint_holder = Arc::new(Mutex::new(None));
+    let (tx, rx) = tokio::sync::oneshot::channel();
+    let cache = EndpointCache {
+        error: error_holder.clone(),
+        endpoint: endpoint_holder.clone(),
+        _drop_guard: Arc::new(tx),
+    };
+    let reloader = ReloadEndpoint {
+        loader: Box::new(move || Box::pin((loader_fn)()) as _),
+        endpoint: endpoint_holder,
+        error: error_holder,
+        rx,
+        sleep,
+        time,
+    };
+    reloader.reload_once().await;
+    // if we didn't successfully get an endpoint, bail out so the client knows
+    // configuration failed to work
+    cache.resolve_endpoint()?;
+    Ok((cache, reloader))
+}
+
+impl EndpointCache {
+    fn resolve_endpoint(&self) -> aws_smithy_http::endpoint::Result {
+        self.endpoint
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|e| e.endpoint.clone())
```

Comment:
Do we need a `map`? Can't we just call `clone` directly?
```suggestion
            .clone()
```

Reply:
this is plucking out the endpoint in the map (not just cloning)